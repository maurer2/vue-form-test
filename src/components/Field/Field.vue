<template>
  <div class="field">
    <label class="label" :class="{ 'label--is-required': isRequired }" :for="id">
      {{ label }}
    </label>

    <template v-if="type === 'email' || type === 'text'">
      <input
        class="input"
        :name="name"
        :id="id"
        :value="value"
        :type="type"
        :placeholder="placeholder"
        :required="isRequired"
        @input="handleInput"
      >
    </template>

    <template v-if="type === 'textarea'">
      <textarea
        class="textarea"
        cols="20"
        rows="10"
        :name="name"
        :id="id"
        :value="value"
        :placeholder="placeholder"
        :required="isRequired"
        @input="handleInput"
      />
    </template>

    <p class="error" v-if="!isValid && userHasInteractedWithForm">
      Please enter a valid value.
    </p>
  </div>
</template>

<script lang="ts">
import { Component, Prop, Vue, Watch } from 'vue-property-decorator';

@Component
export default class Field extends Vue {
  @Prop() private name!: string;
  @Prop() private id!: string;
  @Prop() private label!: string;
  @Prop() private value!: string;
  @Prop() private placeholder!: string;
  @Prop() private isValid!: boolean;
  @Prop() private isRequired!: boolean;
  @Prop() private type!: 'text' | 'email' | 'textarea';

  private userHasInteractedWithForm: boolean = false;

  handleInput(event: Event): void {
    this.userHasInteractedWithForm = true;

    if (!event) {
      return;
    }

    const { target } = event;
    const newValue = (target as HTMLInputElement).value;

    this.$emit('input', newValue);
  }

  static isValidTextField(fieldValue: string): boolean {
    return fieldValue.length > 0;
  }

  static isValidEmailField(fieldValue: string): boolean {
    // https://emailregex.com/
    // eslint-disable-next-line
    const emailRegex = new RegExp(/^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/);
    const isValid = emailRegex.test(fieldValue);

    return isValid;
  }

  @Watch('value')
  handleValueChange(newValue: string): void {
    const {
 id, type, isValid: oldIsValid,
} = this;
    const eventName: string = 'isValidChange';

    if (type === 'textarea') {
      return;
    }

    const newIsValid = (type === 'text')
      ? Field.isValidTextField(newValue)
      : Field.isValidEmailField(newValue);

    if (newIsValid !== oldIsValid) {
      this.$emit(eventName, newIsValid, id);
    }
  }
}
</script>

<style scoped lang="scss">
.field {
  display: flex;
  flex-wrap: wrap;
  align-items: center;

  & + & {
    margin-top: 0.5rem;
  }
}

.label {
  margin: 0.5rem 0;
  flex-basis: 10rem;
  flex-grow: 0;
  align-self: baseline;
  font-weight: bold;

  &--is-required:after {
    content: "*";
  }
}

.input,
.textarea {
  flex-grow: 1;
}

.error {
  margin: 0.5rem 0;
  flex-grow: 1;
  color: $orange;
}
</style>
