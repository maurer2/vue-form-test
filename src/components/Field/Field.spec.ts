import { shallowMount } from '@vue/test-utils';
import Field from 'components/Field/Field.vue';

describe('Field', () => {
  let wrapper: any;

  const mockedInputEvent = jest.fn();

  beforeEach(() => {
    wrapper = shallowMount(Field, {
      propsData: {
        name: 'name',
        id: 'id',
        label: 'label',
        value: 'value',
        placeholder: 'placeholder',
        isValid: true,
        isRequired: true,
        type: 'text',
      },
      attachToDocument: true,
    });
  });

  afterEach(() => {
    wrapper.destroy();
  });

  test('Component matches snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  test('Field contains child elements', () => {
    expect(wrapper.find('.field').exists()).toBe(true);
    expect(wrapper.find('.label').exists()).toBe(true);
    expect(wrapper.find('.input').exists()).toBe(true);
  });

  test('Field label has correct attributes', () => {
    expect(wrapper.find('.label').attributes('for')).toBe('id');
    expect(wrapper.find('.label').classes()).toContain('label');
  });

  test('Field label only has "label--is-required" if required prop is set', () => {
    wrapper.setProps({ isRequired: true });
    expect(wrapper.find('.label').classes()).toContain('label--is-required');

    wrapper.setProps({ isRequired: false });
    expect(wrapper.find('.label').classes()).not.toContain('label--is-required');
  });

  test('Field input has correct attributes', () => {
    expect(wrapper.find('.input').attributes('name')).toBe('name');
    expect(wrapper.find('.input').attributes('id')).toBe('id');
    expect(wrapper.find('.input').attributes('type')).toBe('text');
    expect(wrapper.find('.input').attributes('placeholder')).toBe('placeholder');
    expect(wrapper.find('.input').attributes('required')).toBe('required');
    expect(wrapper.find('.input').classes()).toContain('input');
  });

  test('Field label only has required-attribute if required prop is set', () => {
    wrapper.setProps({ isRequired: true });
    expect(wrapper.find('.input').attributes('required')).toBe('required');

    wrapper.setProps({ isRequired: false });
    expect(wrapper.find('.label').attributes('required')).not.toBe('required');
  });

  test('Field error element is only shown if field is invalid and user has made a previous input', () => {
    wrapper.setData({ userHasInteractedWithForm: true });

    wrapper.setProps({ isValid: true });
    expect(wrapper.find('.error').exists()).toBe(false);

    wrapper.setProps({ isValid: false });
    expect(wrapper.find('.error').exists()).toBe(true);
  });

  test('Field error element is not shown if field is invalid and user has not interacted with field', () => {
    wrapper.setData({ userHasInteractedWithForm: false });

    wrapper.setProps({ isValid: true });
    expect(wrapper.find('.error').exists()).toBe(false);

    wrapper.setProps({ isValid: false });
    expect(wrapper.find('.error').exists()).toBe(false);
  });

  test('Field handleInput method is called on input', () => {
    wrapper.setMethods({ handleInput: mockedInputEvent });
    wrapper.find('.input').trigger('input');

    expect(mockedInputEvent).toBeCalled();
  });

  test('Field handleInput method emits an event to parent', () => {
    wrapper.find('.input').trigger('input');

    expect(wrapper.emitted('input')).toBeTruthy();
  });

  test.skip('Field isValidTextField returns correct value', () => {
    wrapper.vm.isValidTextField('test').toBe(true);
    wrapper.vm.isValidTextField('test test test').toBe(true);
    wrapper.isValidTextField('12345').toBe(true);
    wrapper.isValidTextField('').toBe(false);
  });

  test.skip('Field isValidEmailField returns correct value', () => {
    wrapper.vm.isValidEmailField('test').toBe(false);
    wrapper.vm.isValidEmailField('test test test').toBe(false);
    wrapper.vm.isValidEmailField('12345').toBe(false);
    wrapper.vm.isValidEmailField('').toBe(false);
    wrapper.vm.isValidEmailField('aaa@wefewf.com').toBe(true);
    wrapper.vm.isValidEmailField('aaa.wefwef@wefewf.com').toBe(true);
  });
});
