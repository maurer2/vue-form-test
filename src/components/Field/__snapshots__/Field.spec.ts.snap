// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Field Component matches snapshot 1`] = `
<div class="field"><label for="id" class="label label--is-required">
    label
  </label> <input name="name" id="id" type="text" placeholder="placeholder" required="required" class="input">
  <!---->
  <!---->
</div>
`;
